README.md
# rmexpbak

Delete backups leaving only a very small fraction of them.

Before I upgrade my system I make a backup of the root filesystem in the form of
a Btrfs snapshot in `/media/hdd/backups`. These backups end up taking a lot of
space, so I regularly delete some of them. However, it can be tricky to decide
which backups to delete.

Suppose in `/media/hdd/backups` I have 5 btrfs snapshots named by date.

```
2021-09-25  2021-11-07  2021-12-01  2022-01-15  2022-02-27
```

This program groups backups into increasing powers of two and deletes all but
one backup in the group. This means an exponential number of backups are deleted
and a logarithmic amount of backups are kept. Since so few backups are kept a
lot of disk space is saved.

```
2021-09-25  2021-11-07 |  2021-12-01  2022-01-15  |   2022-02-27      |
 Group of 4 backups    |  Group of 2 backups      | Group of 1 backup |
```

Keep one backup in each group.

```
                       |              2022-01-15  |   2022-02-27      |
 Group of 4 backups    |  Group of 2 backups      | Group of 1 backup |
```

Note the "Group of 4 backups" deletes all the backups because there is not
enough backups to make a real group of 4.

Of the 5 backups in the example above, only floor(lb(5)) = 2 are kept. The
pruning is more drastic if there are more backups. For example, if there were
256 backups, only 8 would be kept. This allows you to make backups everyday
without wasting a lot of space.

The will program remember the backups it deleted, so running the program twice
on the same directory should be a no-op if no additional backups have been made.

## Usage

For the program to work, it needs to know the command used to delete backups.
Create a [Dhall configuration file](https://dhall-lang.org/) named
`.rmexpbak.dhall` in the directory containing the backups.  e.g.

``` dhall
{rmCmd = ["btrfs", "subvolume", "delete"]}
```

Run `rmexpbak <path to backups>` to see the commands the program will run. Once
you are satisfied, run `rmexpbak --delete <path to backups>` to actually delete
the backups. e.g.

``` sh
$ tee /media/hdd/backups/.rmexpbak.dhall << EOF
{rmCmd = ["btrfs", "subvolume", "delete"]}
EOF
$ rmexpbak /media/hdd/backups/
Dry-run. Specify --delete to actually delete files
["btrfs","subvolume","delete","/media/hdd/backups/2021-09-25"]
["btrfs","subvolume","delete","/media/hdd/backups/2021-11-07"]
["btrfs","subvolume","delete","/media/hdd/backups/2021-12-01"]
Dry-run. Specify --delete to actually delete files
$ rmexpbak --delete /media/hdd/backups
["btrfs","subvolume","delete","/media/hdd/backups/2021-09-25"]
Delete subvolume (no-commit): '/media/hdd/backups/2021-09-25'
["btrfs","subvolume","delete","/media/hdd/backups/2021-11-07"]
Delete subvolume (no-commit): '/media/hdd/backups/2021-11-07'
["btrfs","subvolume","delete","/media/hdd/backups/2021-12-01"]
Delete subvolume (no-commit): '/media/hdd/backups/2021-12-01'
```

## Pre-commit checklist

1. Run tests (`stack test`)
2. Run coverage on unit tests (`stack test --coverage`)
3. Run hlint (`hlint .`)



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.0] - 2022-03-13
- Initial release

[1.0.0]: https://github.com/moyamo/rmexpbak/releases/tag/v1.0.0


test/Spec.hs
-- We have orphan BakName instances to make testing easier
{-# OPTIONS_GHC -Wno-orphans #-}

module Main where

import qualified Data.List as List
import qualified Data.Map as Map
import qualified Data.Set as Set
import qualified Data.Tuple as Tuple
import qualified Dhall
import Lib
import RIO
import RIO.Directory as Directory
import RIO.FilePath ((</>))
import System.Environment as Environment
import Test.Hspec
import Test.Hspec.QuickCheck
import Test.QuickCheck
import Test.QuickCheck.Instances.Text ()

instance Arbitrary BakName where
  arbitrary = BakName <$> arbitrary

instance IsString BakName where
  fromString = BakName

main :: IO ()
main = hspec $ do
  describe "serializeDhall" $ do
    let serializeThenParse :: (Dhall.FromDhall a, Dhall.ToDhall a) => a -> IO a
        serializeThenParse =
          Dhall.input Dhall.auto
            . textDisplay
            . serializeDhall
    prop "is right inverse of (Dhall.input . Dhall.auto) " $
      \v -> serializeThenParse (v :: [(String, Int)]) >>= flip shouldBe v

  describe "listBaksInLoft" $ do
    it "returns all files and directories that are not hidden" $ do
      withSystemTempDirectory "rmexpbak_test" $ \tmpDir -> do
        traverse_ (\n -> Directory.createDirectory (tmpDir </> n)) $
          [ "2020-01-03",
            ".hidden_dir"
          ]
        traverse_ (\n -> writeFileUtf8Builder (tmpDir </> n) "") $
          [ "2021-01-06.zip",
            ".rmexpbak_state.dhall",
            "someFile",
            ".rmexpbak.dhall"
          ]
        baks <- listBaksInLoft (LoftDir tmpDir)
        baks `shouldSatisfy` elemsUnique
        Set.fromList baks `shouldBe` Set.fromList ["2020-01-03", "2021-01-06.zip", "someFile"]

  describe "updateBackupNumbers" $ do
    prop "is identity when bns == Map.keys bo" $
      \v -> updateBackupNumbers (Map.keys v) v `shouldBe` v
    it "is empty when empty" $
      updateBackupNumbers [] Map.empty `shouldBe` Map.empty
    it "starts with one" $
      updateBackupNumbers ["2021-01-01"] Map.empty `shouldBe` Map.fromList [("2021-01-01", 1)]
    it "keeps date order" $
      shouldBe
        ( updateBackupNumbers
            ["2021-03-01", "2021-01-01", "2021-01-03"]
            Map.empty
        )
        ( Map.fromList
            [("2021-01-01", 1), ("2021-01-03", 2), ("2021-03-01", 3)]
        )
    it "adds new dates at end" $
      shouldBe
        ( updateBackupNumbers
            ["2021-05-05", "2021-03-01", "2021-01-01"]
            (Map.fromList [("2021-01-01", 1), ("2020-01-03", 2), ("2021-03-01", 3)])
        )
        ( Map.fromList
            [("2021-01-01", 1), ("2020-01-03", 2), ("2021-03-01", 3), ("2021-05-05", 4)]
        )
    it "adds new past dates at end" $
      shouldBe
        ( updateBackupNumbers
            ["2020-05-05", "2021-03-01", "2021-01-01"]
            (Map.fromList [("2021-01-01", 1), ("2020-01-03", 2), ("2021-03-01", 3)])
        )
        ( Map.fromList
            [("2021-01-01", 1), ("2020-01-03", 2), ("2021-03-01", 3), ("2020-05-05", 4)]
        )
    prop "return value is a submap of bo" $
      \(v, w) -> v `Map.isSubmapOf` updateBackupNumbers w v `shouldBe` True
    prop "has unique integers" $
      \(v, w) ->
        elemsUnique (v & Map.elems)
          ==> elemsUnique (updateBackupNumbers w v & Map.elems) `shouldBe` True

  describe "numbersToKeep" $ do
    prop "never returns numbers larger than given" $
      \n -> numbersToKeep n `shouldSatisfy` all (<= n)
    it "returns [0] when given 0" $
      numbersToKeep 0 `shouldBe` [0]
    it "returns [0, 4, 6, 7] when given 7" $
      numbersToKeep 7 `shouldBe` [0, 4, 6, 7]
    it "returns [0, 8, 12, 14, 16] when given 16" $
      numbersToKeep 16 `shouldBe` [0, 8, 12, 14, 16]
    it "returns [0, 8, 12, 16, 17] when given 17" $
      numbersToKeep 17 `shouldBe` [0, 8, 12, 16, 17]
    it "returns [0, 2] when given 2" $
      numbersToKeep 2 `shouldBe` [0, 2]
    prop "always returns n as last value" $
      \n -> (numbersToKeep n & List.last) `shouldBe` n
    prop "always returns 0 as the first value" $
      \n -> (numbersToKeep n & List.head) `shouldBe` 0

  describe "baksToDelete" $ do
    prop "never deletes the latest backup" $ do
      \v ->
        (v & Map.elems & elemsUnique) -- Integers are unique
          ==> baksToDelete v `shouldSatisfy` ((v & Map.toList & fmap Tuple.swap & List.maximum & snd) `notElem`)
    prop "never deletes the 0th backup" $ do
      \v ->
        (v & Map.elems & elemsUnique) -- Integers are unique
          ==> baksToDelete v `shouldSatisfy` ((v & Map.toList & fmap Tuple.swap & Map.fromList & Map.lookup 0 & fromMaybe "trivial case") `notElem`)
    it "deletes for adding 3 correctly" $
      baksToDelete (Map.fromList [("2020-01-03", 2), ("2021-03-01", 3)]) `shouldBe` []
    it "deletes for adding 4 correctly" $
      baksToDelete (Map.fromList [("2020-01-03", 2), ("2021-03-01", 3), ("2022-01-01", 4)]) `shouldBe` ["2021-03-01"]

  describe "rmexpbakMain" $ do
    let withTestDir go = withSystemTempDirectory "rmexpbak_test" $ \tmpDir -> do
          traverse_ (\n -> Directory.createDirectory (tmpDir </> n)) $
            [ "2",
              "4",
              "5",
              "6",
              "7",
              ".hidden_dir"
            ]
          traverse_ (\n -> writeFileUtf8Builder (tmpDir </> n) "") $
            [ "1.zip",
              "3.zip",
              "8.zip",
              "9.zip"
            ]
          writeFileUtf8Builder (tmpDir </> ".rmexpbak.dhall") "{ rmCmd = [\"rm\",\"-r\"] }"
          go tmpDir
    it "doesn't delete anything when --delete not specified" $ do
      withTestDir $ \tmpDir -> do
        Environment.withArgs [tmpDir] $ do
          rmexpbakMain
          remainingBaks <- Directory.listDirectory tmpDir
          remainingBaks `shouldSatisfy` elemsUnique
          Set.fromList remainingBaks
            `shouldBe` Set.fromList
              [ ".rmexpbak.dhall",
                ".rmexpbak_state.dhall",
                ".hidden_dir",
                "1.zip",
                "2",
                "3.zip",
                "4",
                "5",
                "6",
                "7",
                "8.zip",
                "9.zip"
              ]
    it "deletes an exponential number of backups when --delete is specified" $ do
      withTestDir $ \tmpDir -> do
        Environment.withArgs [tmpDir, "--delete"] $ do
          rmexpbakMain
          remainingBaks <- Directory.listDirectory tmpDir
          remainingBaks `shouldSatisfy` elemsUnique
          Set.fromList remainingBaks
            `shouldBe` Set.fromList
              [ ".rmexpbak.dhall",
                ".rmexpbak_state.dhall",
                ".hidden_dir",
                "4",
                "8.zip",
                "9.zip"
              ]
    it "remembers the numbering when run consecutively" $ do
      withTestDir $ \tmpDir -> do
        Environment.withArgs [tmpDir, "--delete"] $ do
          rmexpbakMain
          remainingBaks <- Directory.listDirectory tmpDir
          remainingBaks `shouldSatisfy` elemsUnique
          Set.fromList remainingBaks
            `shouldBe` Set.fromList
              [ ".rmexpbak.dhall",
                ".rmexpbak_state.dhall",
                ".hidden_dir",
                "4",
                "8.zip",
                "9.zip"
              ]
          Directory.createDirectory (tmpDir </> "0.d")
          rmexpbakMain
          remainingBaks2 <- Directory.listDirectory tmpDir
          remainingBaks2 `shouldSatisfy` elemsUnique
          Set.fromList remainingBaks2
            `shouldBe` Set.fromList
              [ ".rmexpbak.dhall",
                ".rmexpbak_state.dhall",
                ".hidden_dir",
                "4",
                "8.zip",
                "0.d"
              ]

elemsUnique :: Ord a => [a] -> Bool
elemsUnique l = (l & length) == (l & Set.fromList & Set.size)


src/Lib.hs
{-# LANGUAGE TemplateHaskell #-}

module Lib where

import qualified Data.List.Index as List (indexed)
import qualified Data.Map as Map
import qualified Data.Maybe as Maybe
import qualified Data.Set as Set
import Data.Tuple (swap)
import Dhall (FromDhall, ToDhall)
import qualified Dhall
import Dhall.Deriving (type (<<<))
import qualified Dhall.Deriving as Dhall
import qualified Dhall.Pretty as Dhall
import Options.Applicative.Simple
  ( ReadM,
    argument,
    empty,
    help,
    long,
    metavar,
    short,
    simpleOptions,
    simpleVersion,
    switch,
  )
import qualified Options.Applicative.Simple as Options.Applicative
import qualified Paths_rmexpbak
import RIO
import qualified RIO.Directory as Directory
import RIO.FilePath ((</>))
import RIO.Lens
import qualified RIO.List as List
import RIO.Process
import qualified RIO.Process as Process
import qualified RIO.Text as Text
import qualified System.IO.Error

-- | 'rmexpbakMain' is the main function called in Main. It is the entry-point
-- of this program.
rmexpbakMain :: IO ()
rmexpbakMain = do
  cwd <- Directory.getCurrentDirectory
  cla <- parseCmdLineArgs cwd
  lo <-
    logOptionsHandle stderr $
      False -- Disable verbose output
  pc <- mkDefaultProcessContext
  withLogFunc lo $ \lf ->
    let app =
          App
            { appLogFunc = lf,
              appProcessContext = pc,
              appCmdLineArgs = cla
            }
     in runRIO app run

-- * RIO Boilerplate

data App = App
  { appLogFunc :: !LogFunc,
    appProcessContext :: !ProcessContext,
    appCmdLineArgs :: !CmdLineArgs
  }

instance HasLogFunc App where
  logFuncL = lens appLogFunc (\x y -> x {appLogFunc = y})

instance HasProcessContext App where
  processContextL = lens appProcessContext (\x y -> x {appProcessContext = y})

-- * Parse the command-line arguments

data CmdLineArgs = CmdLineArgs
  { cmdLineArgsDelete :: Bool,
    cmdLineArgsPath :: LoftDir
  }
  deriving (Show)

class HasCmdLineArgs env where
  cmdLineArgsL :: Lens' env CmdLineArgs

instance HasCmdLineArgs App where
  cmdLineArgsL = lens appCmdLineArgs (\x y -> x {appCmdLineArgs = y})

parseCmdLineArgs :: FilePath -> IO CmdLineArgs
parseCmdLineArgs workingDirectory = do
  (options, ()) <-
    simpleOptions
      $(simpleVersion Paths_rmexpbak.version)
      empty -- Text before "Usage:" line
      "Delete an exponential number of backups"
      ( CmdLineArgs
          <$> switch
            ( long "delete"
                <> short 'd'
                <> help "Actually delete backups. Without this flag we just print the files that would be deleted."
            )
            <*> argument (readLoftDir workingDirectory) (metavar "DIR" <> help "The directory containing the backups")
      )
      empty
  return options
  where
    readLoftDir :: FilePath -> ReadM LoftDir
    readLoftDir cwd = do
      path <- Options.Applicative.str
      return $ LoftDir (cwd </> path)

-- * State

-- A Loft is the folder that contains the backups.

-- | Keeps the state of the Loft across invocations of this program.
data LoftState = LoftState
  { -- | If @x@ is the @n@th backup then @@Map.lookup x loftStateBakOrder ==
    -- Just n@@
    loftStateBakOrder :: Map BakName Word
  }
  deriving (Generic)
  deriving
    (ToDhall, FromDhall)
    via Dhall.Codec
          (Dhall.Field (Dhall.CamelCase <<< Dhall.DropPrefix "loftState"))
          LoftState

{- HLINT ignore LoftState "Use newtype instead of data" -}

-- | The absolute path to the directory that contains the backups
newtype LoftDir = LoftDir {unLoftDir :: FilePath}
  deriving (Show, Eq, Ord, ToDhall, FromDhall) via FilePath

-- | The filename or dirname of the backup. eg. The BakName of
-- @\/var\/backups\/2021-01-01.zip@ is just @2021-01-01.zip@
newtype BakName = BakName {unBakName :: FilePath}
  deriving (Eq, Ord, Show, ToDhall, FromDhall) via FilePath

-- | Takes a LoftDir and converts it into a list of BakName ignoring hidden
-- files
listBaksInLoft :: LoftDir -> IO [BakName]
listBaksInLoft ld = do
  baks <- Directory.listDirectory $ unLoftDir ld
  return $ [BakName b | b <- baks, not ("." `List.isPrefixOf` b)]

bakToPath :: LoftDir -> BakName -> FilePath
bakToPath ld bn = unLoftDir ld </> unBakName bn

-- | Updates the bakOrder with the new Baks found in the path
updateBackupNumbersIO :: LoftDir -> Map BakName Word -> IO (Map BakName Word)
updateBackupNumbersIO ld bo = do
  baks <- listBaksInLoft ld
  -- ignore hidden files
  return $ updateBackupNumbers baks bo

updateBackupNumbers :: [BakName] -> Map BakName Word -> Map BakName Word
updateBackupNumbers bns bo =
  Map.union
    bo
    ( newBaks
        & List.sort
        & indexedStarting nextNumber
        & fmap swap
        & Map.fromList
    )
  where
    newBaks = filter (`Map.notMember` bo) bns
    nextNumber = (bo & Map.elems & foldl' max 0) + 1
    indexedStarting n = (each . _1 %~ (+ n) . fromIntegral) . List.indexed

readDhallIfExists :: FromDhall a => FilePath -> RIO env (Maybe a)
readDhallIfExists path =
  catch
    ( do
        utf8 <- readFileUtf8 path
        Just <$> liftIO (Dhall.input Dhall.auto utf8)
    )
    ( \case
        e
          | System.IO.Error.isDoesNotExistError e ->
            return Nothing
          | otherwise -> throwIO e
    )

-- | Save the new LoftState to disk
updateLoftStateRIO :: HasCmdLineArgs env => RIO env LoftState
updateLoftStateRIO = do
  loftDir <- view (cmdLineArgsL . to cmdLineArgsPath)
  let loftStateName = ".rmexpbak_state.dhall"
  let loftStatePath = unLoftDir loftDir </> loftStateName
  loftState <-
    readDhallIfExists loftStatePath >>= \case
      Nothing -> return $ LoftState {loftStateBakOrder = Map.empty}
      Just ls -> return ls
  let bo = loftStateBakOrder loftState
  newBo <- liftIO $ updateBackupNumbersIO loftDir bo
  let newLoftState = loftState {loftStateBakOrder = newBo}
  writeFileUtf8Builder loftStatePath (serializeDhall newLoftState)
  return newLoftState

serializeDhall :: ToDhall a => a -> Utf8Builder
serializeDhall =
  (<> "\n") . displayShow . Dhall.prettyExpr
    . Dhall.embed Dhall.inject

-- | We need to decide what Baks we want to delete. As is this algorithm named
-- we want to delete an exponential number of backups. We do this by keeping the
-- largest number divisible by 2^n for every n.
numbersToKeep :: Word -> [Word]
numbersToKeep lastNumber = go (Just lastNumber) 1 []
  where
    go Nothing _ acc = acc
    go (Just n) pow2 acc =
      let numbers = [n, n - 1 .. 0]
          r =
            take (fromIntegral pow2) numbers
              & filter (`divisible` pow2)
              & List.headMaybe
              -- A sequence of d consecutive numbers must have at least one
              -- number divisible by d
              & Maybe.fromJust
          -- Suppose we didn't update n, if lastNumber is, say, 16 then 16 is
          -- the largest number divisible by 1, 2, 4, 8 and 16, thus the
          -- algorithm would say keep [16]. However if we subtract pow2 from n
          -- then the algorithm says keep [0, 8, 12, 14, 16], which is much
          -- more useful.
          newN
            | n >= pow2 = Just $ n - pow2
            | otherwise = Nothing
       in go newN (2 * pow2) (r : acc)
    divisible n d = n `rem` d == 0

baksToDelete :: Map BakName Word -> [BakName]
baksToDelete bo = bo & Map.filter (`notElem` keep) & Map.keys
  where
    keep = numbersToKeep maxNumber & Set.fromList
    maxNumber = bo & Map.elems & foldl' max 0

-- * Read a configuration file that tells us how to delete the file

data LoftConfig = LoftConfig
  { loftConfigRmCmd :: [Text]
  }
  deriving (Generic)
  deriving
    (FromDhall)
    via Dhall.Codec
          (Dhall.Field (Dhall.CamelCase <<< Dhall.DropPrefix "loftConfig"))
          LoftConfig

{- HLINT ignore LoftConfig "Use newtype instead of data" -}

readLoftConfig :: HasCmdLineArgs env => RIO env LoftConfig
readLoftConfig = do
  loftDir <- view (cmdLineArgsL . to cmdLineArgsPath)
  let loftConfigName = ".rmexpbak.dhall"
  let loftConfigPath = unLoftDir loftDir </> loftConfigName
  readDhallIfExists loftConfigPath >>= \case
    Just lc -> return lc
    Nothing ->
      liftIO $
        fail $
          "Could not find .rmexpbak.dhall configuration file in "
            <> unLoftDir loftDir

deleteBaks :: LoftConfig -> LoftState -> RIO App ()
deleteBaks loftConfig loftState = do
  let toDelete = baksToDelete $ loftStateBakOrder loftState
  let rmCmd = Text.unpack <$> loftConfigRmCmd loftConfig
  shouldDelete <- view (cmdLineArgsL . to cmdLineArgsDelete)
  loftDir <- view (cmdLineArgsL . to cmdLineArgsPath)
  toDelete
    & traverse_
      ( \bak -> do
          let bakPath = loftDir `bakToPath` bak
          exists <- liftIO $ Directory.doesPathExist bakPath
          when exists $ do
            logInfo $ displayShow (rmCmd ++ [bakPath])
            when shouldDelete $ do
              let cmd : args = rmCmd ++ [bakPath]
              Process.proc cmd args Process.runProcess_
      )

-- * Delete an Exponential amount of Backups

run :: RIO App ()
run = do
  loftState <- updateLoftStateRIO
  loftConfig <- readLoftConfig
  willDelete <- view (cmdLineArgsL . to cmdLineArgsDelete)
  unless willDelete $ do
    logInfo "Dry-run. Specify --delete to actually delete files"
  deleteBaks loftConfig loftState
  unless willDelete $ do
    logInfo "Dry-run. Specify --delete to actually delete files"
